File: src/programmers/ImportProgrammer.ts
import ts from "typescript";

import { MapUtil } from "../utils/MapUtil";

export class ImportProgrammer {
  private readonly assets_: Map<string, IAsset> = new Map();
  private readonly options_: Readonly<ImportProgrammer.IOptions>;

  public constructor(options?: Partial<ImportProgrammer.IOptions>) {
    this.options_ = {
      internalPrefix: options?.internalPrefix ?? "",
    };
  }

  /* -----------------------------------------------------------
    ENROLLMENTS
  ----------------------------------------------------------- */
  public default(props: ImportProgrammer.IDefault): ts.Identifier {
    const asset: IAsset = this.take(props.file);
    asset.default ??= props;
    asset.default.type ||= props.type;
    return ts.factory.createIdentifier(asset.default.name);
  }

  public instance(props: ImportProgrammer.IInstance): ts.Identifier {
    const alias: string = props.alias ?? props.name;
    const asset: IAsset = this.take(props.file);
    MapUtil.take(asset.instances, alias, () => props);
    return ts.factory.createIdentifier(alias);
  }

  public namespace(props: ImportProgrammer.INamespace): ts.Identifier {
    const asset: IAsset = this.take(props.file);
    asset.namespace ??= props;
    return ts.factory.createIdentifier(asset.namespace.name);
  }

  public type(props: {
    file: string;
    name: string | ts.EntityName;
    arguments?: ts.TypeNode[];
  }): ts.ImportTypeNode {
    return ts.factory.createImportTypeNode(
      ts.factory.createLiteralTypeNode(
        ts.factory.createStringLiteral(props.file),
      ),
      undefined,
      typeof props.name === "string"
        ? ts.factory.createIdentifier(props.name)
        : props.name,
      props.arguments,
    );
  }

  /**
   * @internal
   */
  public internal(name: string): ts.PropertyAccessExpression {
    if (name.startsWith("_") === false) name = `_${name}`;
    return ts.factory.createPropertyAccessExpression(
      this.namespace({
        file: `typia/lib/internal/${name}.js`,
        name: this.alias(name),
      }),
      name,
    );
  }

  /**
   * @internal
   */
  public getInternalText(name: string): string {
    if (name.startsWith("_") === false) name = `_${name}`;
    const asset: IAsset | undefined = this.take(
      `typia/lib/internal/${name}.js`,
    );
    if (!asset?.namespace) throw new Error(`Internal asset not found: ${name}`);
    return `${asset.namespace.name}.${name}`;
  }

  /**
   * @internal
   */
  private take(file: string): IAsset {
    return MapUtil.take(this.assets_, file, () => ({
      file,
      default: null,
      namespace: null,
      instances: new Map(),
    }));
  }

  private alias(name: string): string {
    return `__${this.options_.internalPrefix}${name}`;
  }

  /* -----------------------------------------------------------
    PROGRAM STATEMENTS
  ----------------------------------------------------------- */
  public toStatements(): ts.ImportDeclaration[] {
    const statements: ts.ImportDeclaration[] = [];
    for (const asset of this.assets_.values()) {
      if (asset.namespace !== null)
        statements.push(
          ts.factory.createImportDeclaration(
            undefined,
            ts.factory.createImportClause(
              false,
              undefined,
              ts.factory.createNamespaceImport(
                ts.factory.createIdentifier(asset.namespace.name),
              ),
            ),
            ts.factory.createStringLiteral(asset.file),
          ),
        );
      if (asset.default !== null)
        statements.push(
          ts.factory.createImportDeclaration(
            undefined,
            ts.factory.createImportClause(
              asset.default.type,
              ts.factory.createIdentifier(asset.default.name),
              undefined,
            ),
            ts.factory.createStringLiteral(asset.file),
          ),
        );
      if (asset.instances.size > 0)
        statements.push(
          ts.factory.createImportDeclaration(
            undefined,
            ts.factory.createImportClause(
              false,
              undefined,
              asset.instances.size > 0
                ? ts.factory.createNamedImports(
                    [...asset.instances.values()].map((ins) =>
                      ts.factory.createImportSpecifier(
                        false,
                        ins.alias || ins.alias === ins.name
                          ? ts.factory.createIdentifier(ins.name)
                          : undefined,
                        ts.factory.createIdentifier(ins.alias ?? ins.name),
                      ),
                    ),
                  )
                : undefined,
            ),
            ts.factory.createStringLiteral(asset.file),
            undefined,
          ),
        );
    }
    return statements;
  }
}

export namespace ImportProgrammer {
  export interface IOptions {
    internalPrefix: string;
  }

  export interface IDefault {
    file: string;
    name: string;
    type: boolean;
  }
  export interface IInstance {
    file: string;
    name: string;
    alias: string | null;
  }
  export interface INamespace {
    file: string;
    name: string;
  }
}

interface IAsset {
  file: string;
  default: ImportProgrammer.IDefault | null;
  namespace: ImportProgrammer.INamespace | null;
  instances: Map<string, ImportProgrammer.IInstance>;
}

File: packages/core/src/programmers/ImportProgrammer.go
package programmers

import "strings"

type ImportProgrammer struct {
	assets  map[string]*importProgrammerAsset
	options ImportProgrammerOptions
}

type ImportProgrammerOptions struct {
	InternalPrefix string
	Runtime        *string
}

type ImportProgrammerDefault struct {
	File string
	Name string
	Type bool
}

type ImportProgrammerInstance struct {
	File  string
	Name  string
	Alias *string
}

type ImportProgrammerNamespace struct {
	File string
	Name string
}

type importProgrammerAsset struct {
	File      string
	Default   *ImportProgrammerDefault
	Namespace *ImportProgrammerNamespace
	Instances map[string]ImportProgrammerInstance
}

func NewImportProgrammer(options *ImportProgrammerOptions) *ImportProgrammer {
	next := ImportProgrammerOptions{}
	if options != nil {
		next = *options
	}
	return &ImportProgrammer{
		assets:  map[string]*importProgrammerAsset{},
		options: next,
	}
}

func (importer *ImportProgrammer) Default(props ImportProgrammerDefault) string {
	asset := importer.take(props.File)
	if asset.Default == nil {
		copied := props
		asset.Default = &copied
	}
	if asset.Default.Type == false {
		asset.Default.Type = props.Type
	}
	return asset.Default.Name
}

func (importer *ImportProgrammer) Instance(props ImportProgrammerInstance) string {
	alias := props.Name
	if props.Alias != nil {
		alias = *props.Alias
	}
	asset := importer.take(props.File)
	if _, ok := asset.Instances[alias]; !ok {
		asset.Instances[alias] = props
	}
	return alias
}

func (importer *ImportProgrammer) Namespace(props ImportProgrammerNamespace) string {
	asset := importer.take(props.File)
	if asset.Namespace == nil {
		copied := props
		asset.Namespace = &copied
	}
	return asset.Namespace.Name
}

func (importer *ImportProgrammer) Type(props struct {
	File      string
	Name      string
	Arguments []string
}) string {
	if len(props.Arguments) == 0 {
		return `import("` + props.File + `").` + props.Name
	}
	return `import("` + props.File + `").` + props.Name + "<" + strings.Join(props.Arguments, ", ") + ">"
}

func (importer *ImportProgrammer) Internal(name string) string {
	if strings.HasPrefix(name, "_") == false {
		name = "_" + name
	}
	namespace := importer.Namespace(ImportProgrammerNamespace{
		File: "typia/lib/internal/" + name,
		Name: importer.alias(name),
	})
	return namespace + "." + name
}

func (importer *ImportProgrammer) GetInternalText(name string) string {
	if strings.HasPrefix(name, "_") == false {
		name = "_" + name
	}
	asset := importer.take("typia/lib/internal/" + name)
	if asset.Namespace == nil {
		panic("Internal asset not found: " + name)
	}
	return asset.Namespace.Name + "." + name
}

func (importer *ImportProgrammer) ToStatements() []string {
	statements := []string{}
	for _, asset := range importer.assets {
		if asset.Namespace != nil {
			statements = append(statements, `import * as `+asset.Namespace.Name+` from "`+asset.File+`";`)
		}
		if asset.Default != nil {
			prefix := ""
			if asset.Default.Type {
				prefix = "type "
			}
			statements = append(statements, `import `+prefix+asset.Default.Name+` from "`+asset.File+`";`)
		}
		if len(asset.Instances) != 0 {
			instances := make([]string, 0, len(asset.Instances))
			for _, ins := range asset.Instances {
				alias := ins.Name
				if ins.Alias != nil {
					alias = *ins.Alias
				}
				if ins.Alias != nil && *ins.Alias != ins.Name {
					instances = append(instances, ins.Name+" as "+alias)
				} else {
					instances = append(instances, alias)
				}
			}
			statements = append(statements, `import { `+strings.Join(instances, ", ")+` } from "`+asset.File+`";`)
		}
	}
	return statements
}

func (importer *ImportProgrammer) take(file string) *importProgrammerAsset {
	if importer.assets == nil {
		importer.assets = map[string]*importProgrammerAsset{}
	}
	if asset, ok := importer.assets[file]; ok {
		return asset
	}
	asset := &importProgrammerAsset{
		File:      file,
		Instances: map[string]ImportProgrammerInstance{},
	}
	importer.assets[file] = asset
	return asset
}

func (importer *ImportProgrammer) alias(name string) string {
	return "__" + importer.options.InternalPrefix + name
}
